File: solutions/typescript/triangle/1/triangle.ts
export class Triangle {
  private _equilateral = false
  private _isosceles = false
  private _scalene = false

  constructor(...sides: number[]) {
    const [a, b, c] = sides
    const allSidePositive = sides.every(side => side > 0)
    const triangleInequality = a + b > c && a + c > b && b + c > a
    // NOTE: another way of doing it
    // const triangleInequality = Math.max(a, b, c) * 2 < sides.reduce((acc, a) => acc + a, 0)

    if (allSidePositive && triangleInequality) {
      this._equilateral = a === b && b === c
      this._isosceles = a === b || b === c || c === a
      this._scalene = !this._isosceles
    }
  }

  get isEquilateral() {
    return this._equilateral
  }

  get isIsosceles() {
    return this._isosceles
  }

  get isScalene() {
    return this._scalene
  }
}

File: solutions/python/triangle/2/triangle.py
def equilateral(sides: list[int]) -> bool:
    a, b, c = sides
    return triangle_inequality(sides) and a == b and b == c


def isosceles(sides: list[int]) -> bool:
    a, b, c = sides
    return triangle_inequality(sides) and (a == b or b == c or c == a)


def scalene(sides: list[int]) -> bool:
    a, b, c = sides
    return triangle_inequality(sides) and not (a == b or b == c or c == a)


def triangle_inequality(sides: list[int]) -> bool:
    a, b, c = sorted(sides)
    return a + b > c
